[김도연] Mission 3: 회원가입 기능 구현

kdylsky opened the pull request:
>wecode PR message template 및 체크 리스트 입니다. 
아래 사항들을 전부 작성/체크 하시고 PR 해주세요!

## 수정 사항 간략한 한줄 요약
(여기에 수정 사항에 대한 간략한 한줄 요약/제목 작성해 주세요.)
- 회원가입 기능 구현		 

## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- EDIT       : westagram/urls.py
- EDIT       : users/urls.py
- CREATE : users/views.py
- CREATE : users/views.py/check_regx() 


## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)
- 없습니다.

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
조건을 만족한다면 각 빈칸에 x 표시 해주세요
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

Commit:
- Create : UserView
- Edit   : urls.py

Commit:
- Modeify : views.py

Commit:
- Modeify : westagram views.py
	    users views.py

Commit:
-EDIT 	: users/views.py
-CREATE : check_regx()

Commit:
- EDIT : users/views.py

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,52 @@
+                email        = email,
+                password     = password,
+                phone_number = phone_number,
+                create_at    = timezone.now(),
+                update_at    = timezone.now(),
```
```suggestion
```
created_at 과 updated_at은 필드 특성상 자동으로 생성될 것입니다.

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,52 @@
+
+from django.http import JsonResponse
+from django.views import View
+from django.utils import timezone
```
```suggestion
```
따라서 얘도 지워주세요.

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,52 @@
+        except ValueError as e:
+            return JsonResponse({"message": f"{e}"}, status=400)
+
+        else:
+            user = User.objects.create(
```
```suggestion
        User.objects.create(
```
else: 는 애초에 불필요해보입니다.
추가로 create 한 결과를 user에 저장할 필요가 없어보입니다. user 객체가 재활용되지 않기 때문입니다!

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,52 @@
+            return JsonResponse({"message": "SUCCESS"}, status=201)
+
+def check_regx(pattern, data):
+    if re.compile(pattern).match(data) is None:
+        raise ValueError("INVALID_PASSWORD")
+        return False
+    else:
+        return True
```
```suggestion
    if not re.compile(pattern).match(data):
        raise ValueError("INVALID_PASSWORD")
```
true or false를 굳이 반환해줄 필요없이 그냥 문제가 있으면 에러를 raise 하면 될 것 같네요!
추가로 truthy or falsy한 값을 비교할 때는 not을 사용할 수 있습니다.

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,52 @@
+            regx_email      = '^[a-zA-Z0-9+-_.]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$'
+            regx_password   = '^(?=.*[A-Za-z])(?=.*\d)(?=.*[$@$!%*#?&])[A-Za-z\d$@$!%*#?&]{8,}$'
+
+            if check_regx(regx_email, email):
+                if User.objects.filter(email=email).exists():
+                    raise ValueError("EXISTED E_MAIL")
+            if check_regx(regx_password, password):
+                if User.objects.filter(phone_number=phone_number).exists():
+                    raise ValueError("EXISTED PHONE_NUMBER")
```
이 부분도 그에 맞게 한 번 수정해보세요!

kdylsky replied:
감사합니다. 수정 후 다시 요청드리겠습니다.

kdylsky replied:
알려주신 Customer Manager는 따로 구현해 보겠습니다.

bigfanoftim reviewed (changes requested):
도연님!!! 하나의 기능도 디테일하게 다뤄보는 것 너무 보기 좋습니다!!
그에 맞게 리뷰를 남겨드렸으니 한번 수정해보세요~

Commit:
- EDIT : users/views.py
- EDIT : users/views.py check_regx()

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,44 @@
+
+def check_regx(pattern, data):
+    if not re.compile(pattern).match(data):
+        raise ValueError("INVALID_PASSWORD")
```
이렇게 하면 모든 예외 케이스에 대해 하나의 에러 메세지만 나오지 않나요?

kdylsky replied:
모든 케이스는 아니지만 이메일 형식을 확인 하는 부분에서 password가 잘못됬다는 에러가 발생하는 것을 확인 했습니다.

확인 후 다시 요청드리겠습니다.

kdylsky replied:
def check_regx(pattern, data):
    if not re.compile(pattern).match(data):
        return True

함수를 구현 후 return을 하지 않을 경우라면 각각의 요소에 해당하는 정규식 확인 함수를 아래와 같이 따로 만들어 주어야 하는 건가요?

def check_email_regx(pattern, data):
    if not re.compile(pattern).match(data):
        raise ValueError("INVALID_EMAIL")

def check_password_regx(pattern, data):
    if not re.compile(pattern).match(data):
        raise ValueError("INVALID_PASSWORD")

kdylsky replied:
1. 핸드폰 유닉크 설정을 위한 핸드폰 정규식 체크 코드를 추가했습니다.

2. 에러메세지 표기를 위해 check_regx(pattern, field_data, target_field)함수에 target_field를 추가해 주었습니다.

3. 말씀하신 에러메시지 표기에 대해서는 제가 아는 한에서 확인한 결과 에러발생에 맞게 에러메시지가 표기 되었습니다.

kdylsky replied:
질문드릴게 있습니다.

코드를 간단히 하기 위해서 함수를 이용했는데, 구현 결과 위와 같이 더 복잡해진 느낌이 든 경우라면 수정해 주어야 하는 것인가요?

bigfanoftim replied:
> 모든 케이스는 아니지만 이메일 형식을 확인 하는 부분에서 password가 잘못됬다는 에러가 발생하는 것을 확인 했습니다.

제가 이전에 말씀드린게 이 부분을 의미한게 맞습니다.

---

더 복잡해진 느낌이라면 어떤 느낌일까요? 그리고 말씀하시는 함수가 정확히 어느 함수인지를 모르겠습니다. 너무 함수가 많아서...
커스텀 매니저에서 사용한 함수를 말씀하시는건지 정규표현식 함수를 말씀하시는건지 정확히 말씀해주세요!

Commit:
- EDIT   : views.py
- EDIT   : check_regx(pattern, field_data) ===> check_regx(pattern, field_data, target_field)
- CREATE : viwes.py/ check phone_number regx
- TEST   : ERROR_TEST

Commit:
- CREATE : custom manager/is_existed() ----> User.objects.is_existed(data)
- EDIT	 : users/models.py
- CREATE : does_existed(data, target_field)
- EDIT	 : users/views.py

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,49 @@
+        except ValueError as e:
+            return JsonResponse({"message": f"{e}"}, status=400)
+
+def check_regx(pattern, field_data, target_field):
+    if not re.compile(pattern).match(field_data):
+        raise ValueError(f"INVILD_{target_field}")
```
에러명을 핸들링하기 위해 `target_field` 인자를 넘겨주신것 같은데
이렇게 모듈을 만들고 사용하면 EMAIL, PASSWORD 부분에서 오타가 날 확률이 있지 않나요?

차라리 check_email_regex(), check_password_regex()로 나누는게 더 깔끔하고 실수를 줄일 수 있을 것 같습니다.

bigfanoftim commented on students/doyeon/users/models.py:
```diff
@@ -1,12 +1,26 @@
 from django.db import models
+from django.db.models import Q
+
+class UserQuerySet(models.QuerySet):
+    def is_existed(self, data):
+        return self.filter(Q(email=data)|Q(phone_number=data))
+
+class UserManager(models.Manager):
+    def get_queryset(self):
+        return UserQuerySet(self.model, using=self._db)
+
+    def is_existed(self, data):
+        return self.get_queryset().is_existed(data)
```
이 코드들은 사용하지 않는 것 같으니 지워주세요.

bigfanoftim commented on students/doyeon/users/views.py:
```diff
@@ -1,3 +1,49 @@
+    if not re.compile(pattern).match(field_data):
+        raise ValueError(f"INVILD_{target_field}")
+
+def does_existed(data, target_field):
+    if User.objects.is_existed(data).exists():
+        raise ValueError(f"EXISTED_{target_field}")
```
이 부분도 따로 커스텀 쿼리셋 만들어 관리할게 아니라면 
여러 함수로 분리하는 것이 더 좋아보입니다.

kdylsky replied:
네 감사합니다. 
커스텀 쿼리셋을 만들면, 만들지 않을때 보다 관리하기가 어려울것 같습니다.
does_existed()함수로 만드는 것이 아닌, 
if 문을 이용해 바로 확인 후 에러코드를 발생시키는코드로 변경하겠습니다.

Commit:
- EDIT	 : models.py
- DELETE : custom manage, custom queryset
- EDIT   : seprate regx_function

bigfanoftim reviewed (approved):
LGTM 👍
